Match more possible patterns for process constructors

klarasm opened the pull request:
While looking at handling exceptions in the LSP, I figured that it might be good to just have the compiler crash less. I started looking at the eta-reduced cases which turned out to not be too hard to fix. Basically, the eta-reduced applications have no Lams in the beginning and then an equal number of Apps removed. Just strip preceeding Lams and an equal number of Apps before matching with the eta-reduced cases instead.

This also attempts to lift local functions in where-scopes to global functions. This might not be the best approach, but seemed to be the least amount of work. Ideally, we should maybe instead change ForSyDeIR to have the functions embedded directly. This would also make it easier when functions are defined globally since we don't need to link them together in CoreIRToProceduralIR. We could then also just ignore any unused global functions.

This also changes the noInlineTypecheck to put NOINLINE on all functions definitions, not just the ones in system.

With a few other changes, this PR also now supports inferred types (but this probably needs more testing). The major remaining restriction after this is that actor definitions still cannot be inlined in the system net list.

klarasm committed:
CoreIRToForSyDeIR: match more possible patterns for processes

This adds two helper functions: stripLam, stripApps.

The first, stripLams, removes all Lams and keeps count of how many where
removed (the count excludes type variables).

The second, stripApps, tries to strip a given number of Apps (returns
Nothing if it couldn't). This means we can only care about the
eta-reduced applications of the processes, and the other (with varying
numbers of signals specified) will be handled automatically.

The third, stripLets, removes internal Let binds inside functions and
lifts them to our global function list. For this to work, they currently
cannot be inlined. Thus this also adds NOINLINE to all functions instead
of just system.

To put it together, stripLams is called first, then remove Let binds and
lift local functions to global functions. This also need all top-level
fuctions to not be inlined, therefore modify noInlineTypecheck for that.
Lastly stripApps is called.

klarasm committed:
CoreIRToProceduralIR: allow unused functions

The user might have defined e.g. test signals which will currently make
the compiler error out. Instead we can ignore these and just don't put
them in the output. This also removes the reduntant listToMaybe

sthaeron commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -54,38 +54,38 @@ initialTranslationContext dflags =
 -- | Translates an `IRFunction` into a Procedural IR function declaration and
 -- optional definition.
-translateIRFunction :: IRFunction -> DynFlags -> [IRConstructor] -> (Global, Maybe Global)
+translateIRFunction :: IRFunction -> DynFlags -> [IRConstructor] -> (Maybe Global, Maybe Global)
```
If you want to have functions in procedural ir be optional rather than erroring out, I would suggest using the type `Maybe (Global, Maybe Global)` as it makes it clear that there could never be a function definition without a function declaration.

sthaeron commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -54,38 +54,38 @@ initialTranslationContext dflags =
         | functionId == targetFunctionId = Just actor
         | otherwise = Nothing
       checkIRConstructor _ = Nothing
-   in listToMaybe (mapMaybe checkIRConstructor constructors)
```
Not sure why this change had to be made, since you are still pattern-matching on the first element in the array of maybes.

klarasm replied:
Yes, this essentially does the same thing, but this turned out shorter as we already do the pattern match and eliminates the need for an import.

sthaeron commented on src/Utilities.hs:
```diff
@@ -103,10 +103,7 @@ noInlineTypecheck tcg = tcg {tcg_binds = applySystemBinds (tcg_binds tcg)}
-           in case funName of
-                "system" -> bind {fun_id = var, fun_matches = applyLocalBinds matches}
-                _ -> bind
+          bind {fun_id = var, fun_matches = applyLocalBinds matches}
```
Why are we now non-inlining all top-level core binds? Is it for `where` scopes that have functions defined within them?

klarasm replied:
This exactly. I originally had it in a separate commit but forgot to add the explanation when i squashed them.

sthaeron commented on src/CoreIRToForSyDeIR.hs:
```diff
@@ -349,6 +351,26 @@ getSourceFromArgument context argument =
 
+-- Strip all Lams so we don't need to bother with non-eta-reduced processes
+stripLamApps :: Integer -> Expr b -> Maybe (Expr b)
+stripLamApps n = \case
```
I would suggest not using `LambdaCase` since we haven't been using it in the rest of the compiler pipeline translation code. It's probably best to keep things consistent.

klarasm replied:
It just feels more natural for me to write it this way, but if you really want to I can do a regular case.

sthaeron commented:
The current implementation has the consequence of being able to compile programs which use eta reduce without a type, but not those that do not use eta reduce and do not have a type:

Meaning we can compile the following program (no type, using eta reduce):
```haskell
module SDF_example_028 where

import ForSyDe.Shallow

-- Netlist
system :: Signal Int -> Signal Int -> Signal Int
system s_in_x s_in_y = s_out
  where
    s_1 = a_a s_in_x s_in_y
    (s_out, s_2) = a_b s_1 s_2_delayed
    s_2_delayed = d_1 s_2

-- Process specifications
a_a = actor21SDF (1, 1) 1 add

a_b = actor22SDF (1, 1) (1, 1) accumulate

d_1 = delaySDF [0]

-- Function definitions
add [x] [y] = [x + y]

accumulate [x] [y] = ([x + y], [x + y])
```

But cannot compile the following example (no type, not using eta reduce):
```haskell
module SDF_example_028 where

import ForSyDe.Shallow

-- Netlist
system :: Signal Int -> Signal Int -> Signal Int
system s_in_x s_in_y = s_out
  where
    s_1 = a_a s_in_x s_in_y
    (s_out, s_2) = a_b s_1 s_2_delayed
    s_2_delayed = d_1 s_2

-- Process specifications
a_a s_1 s_2 = actor21SDF (1, 1) 1 add s_1 s_2

a_b s_1 s_2 = actor22SDF (1, 1) (1, 1) accumulate s_1 s_2

d_1 s = delaySDF [0] s

-- Function definitions
add [x] [y] = [x + y]

accumulate [x] [y] = ([x + y], [x + y])
```
This is mainly a consequence of stripping `App`s in equal amounts to stripping `Lam`s.

sthaeron reviewed (changes requested):
It is very ambitious to make such fundamental changes to our compiler. The stripping Lambdas is quite a smart solution to attempt supporting eta reduced and non-eta reduced haskell functions. Nevertheless, I think the current implementation has some unintended consequences. We should not be stripping `App`s and just have dedicated pattern matching for non-eta reduced and eta reduced cases or update the stripping `Lam`s to identify which inputs are types vs signals.

sthaeron commented:
I've got some ideas of my own for fixing and updating the current implementation in this PR. I'll make some individual commits on your branch. Feel free to rebase them however you'd like.

klarasm commented:
> It is very ambitious to make such fundamental changes to our compiler. The stripping Lambdas is quite a smart solution to attempt supporting eta reduced and non-eta reduced haskell functions. Nevertheless, I think the current implementation has some unintended consequences. We should not be stripping `App`s and just have dedicated pattern matching for non-eta reduced and eta reduced cases or update the stripping `Lam`s to identify which inputs are types vs signals.

I originally added the patterns separately, but it was kind of awful (4 input processes need 4 additional matches and so on, urgh!). I can dig up the original commits but those won't work on definitions without a type either. The type coercion (not sure about the term) will happen inside the function instead which means it won't compile without further modifications. I tried a solution which did this but just stripping the Lams from the types is not enough due to this. When trying further it produced the wrong output so I just left it out.

klarasm commented:
> The current implementation has the consequence of being able to compile programs which use eta reduce without a type, but not those that do not use eta reduce and do not have a type:
> 
> Meaning we can compile the following program (no type, using eta reduce):
> 
> ```haskell
> module SDF_example_028 where
> 
> import ForSyDe.Shallow
> 
> -- Netlist
> system :: Signal Int -> Signal Int -> Signal Int
> system s_in_x s_in_y = s_out
>   where
>     s_1 = a_a s_in_x s_in_y
>     (s_out, s_2) = a_b s_1 s_2_delayed
>     s_2_delayed = d_1 s_2
> 
> -- Process specifications
> a_a = actor21SDF (1, 1) 1 add
> 
> a_b = actor22SDF (1, 1) (1, 1) accumulate
> 
> d_1 = delaySDF [0]
> 
> -- Function definitions
> add [x] [y] = [x + y]
> 
> accumulate [x] [y] = ([x + y], [x + y])
> ```
> 

Hm!

I didn't notice that leaving out the signatures for both the functions and the processes worked, but I guess they might get inferred from the type of the system or something. Either way a nice surprise.

> But cannot compile the following example (no type, not using eta reduce):
> 
> ```haskell
> module SDF_example_028 where
> 
> import ForSyDe.Shallow
> 
> -- Netlist
> system :: Signal Int -> Signal Int -> Signal Int
> system s_in_x s_in_y = s_out
>   where
>     s_1 = a_a s_in_x s_in_y
>     (s_out, s_2) = a_b s_1 s_2_delayed
>     s_2_delayed = d_1 s_2
> 
> -- Process specifications
> a_a s_1 s_2 = actor21SDF (1, 1) 1 add s_1 s_2
> 
> a_b s_1 s_2 = actor22SDF (1, 1) (1, 1) accumulate s_1 s_2
> 
> d_1 s = delaySDF [0] s
> 
> -- Function definitions
> add [x] [y] = [x + y]
> 
> accumulate [x] [y] = ([x + y], [x + y])
> ```
> 
> This is mainly a consequence of stripping `App`s in equal amounts to stripping `Lam`s.

klarasm commented:
> > The current implementation has the consequence of being able to compile programs which use eta reduce without a type, but not those that do not use eta reduce and do not have a type:
> > Meaning we can compile the following program (no type, using eta reduce):
> > ```haskell
> > module SDF_example_028 where
> > 
> > import ForSyDe.Shallow
> > 
> > -- Netlist
> > system :: Signal Int -> Signal Int -> Signal Int
> > system s_in_x s_in_y = s_out
> >   where
> >     s_1 = a_a s_in_x s_in_y
> >     (s_out, s_2) = a_b s_1 s_2_delayed
> >     s_2_delayed = d_1 s_2
> > 
> > -- Process specifications
> > a_a = actor21SDF (1, 1) 1 add
> > 
> > a_b = actor22SDF (1, 1) (1, 1) accumulate
> > 
> > d_1 = delaySDF [0]
> > 
> > -- Function definitions
> > add [x] [y] = [x + y]
> > 
> > accumulate [x] [y] = ([x + y], [x + y])
> > ```
> 
> Hm!
> 
> I didn't notice that leaving out the signatures for both the functions and the processes worked, but I guess they might get inferred from the type of the system or something. Either way a nice surprise.
> 

Tried this example and for me it doesn't actually compile (the type coercions are still in the functions). But we can get to ForSyDeIR at least which is nice for the LSP.

klarasm commented:
If you're interested in the original commit which just adds all the matches:

<details>
<summary>Eta-reduce all matches</summary>
<pre>
diff --git a/src/CoreIRToForSyDeIR.hs b/src/CoreIRToForSyDeIR.hs
index e32f5d2095d1..389e1106666a 100644
--- a/src/CoreIRToForSyDeIR.hs
+++ b/src/CoreIRToForSyDeIR.hs
@@ -358,38 +358,178 @@ getSourceFromArgument context argument =
 -- translation to support additional process constructors.
 translateCoreExpr :: TranslationContext -> CoreBndr -> CoreExpr -> TranslationContext
 translateCoreExpr context binder expr = case expr of
+  -- 0 signals
+  App (App (Var i) _) _
+    | IRVar i == IRString "delaySDF" -> createDelaySDF context binder expr
+  -- 1 signal
   Lam _ (App (App (App (Var i) _) _) _)
     | IRVar i == IRString "delaySDF" -> createDelaySDF context binder expr
+  -- 0 signals
+  App (App (App (App (App (Var i) _) _) _) _) _
+    | IRVar i == IRString "actor11SDF" -> createActorSDF context Actor11 binder expr
+  -- 1 signal
   Lam _ (App (App (App (App (App (App (Var i) _) _) _) _) _) _)
     | IRVar i == IRString "actor11SDF" -> createActorSDF context Actor11 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (Var i) _) _) _) _) _) _
+    | IRVar i == IRString "actor12SDF" -> createActorSDF context Actor12 binder expr
+  -- 1 signal
   Lam _ (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _)
     | IRVar i == IRString "actor12SDF" -> createActorSDF context Actor12 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor13SDF" -> createActorSDF context Actor13 binder expr
+  -- 1 signal
   Lam _ (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _)
     | IRVar i == IRString "actor13SDF" -> createActorSDF context Actor13 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor14SDF" -> createActorSDF context Actor14 binder expr
+  -- 1 signal
   Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _)
     | IRVar i == IRString "actor14SDF" -> createActorSDF context Actor14 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (Var i) _) _) _) _) _) _
+    | IRVar i == IRString "actor21SDF" -> createActorSDF context Actor21 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor21SDF" -> createActorSDF context Actor21 binder expr
+  -- 2 signals
   Lam _ (Lam _ (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _))
     | IRVar i == IRString "actor21SDF" -> createActorSDF context Actor21 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor22SDF" -> createActorSDF context Actor22 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor22SDF" -> createActorSDF context Actor22 binder expr
+  -- 2 signals
   Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _))
     | IRVar i == IRString "actor22SDF" -> createActorSDF context Actor22 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor23SDF" -> createActorSDF context Actor23 binder expr
+  -- 1 signals
+  Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor23SDF" -> createActorSDF context Actor23 binder expr
+  -- 2 signals
   Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _))
     | IRVar i == IRString "actor23SDF" -> createActorSDF context Actor23 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor24SDF" -> createActorSDF context Actor24 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor24SDF" -> createActorSDF context Actor24 binder expr
+  -- 2 signals
   Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _))
     | IRVar i == IRString "actor24SDF" -> createActorSDF context Actor24 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor31SDF" -> createActorSDF context Actor31 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor31SDF" -> createActorSDF context Actor31 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor31SDF" -> createActorSDF context Actor31 binder expr
+  -- 3 signals
   Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _)))
     | IRVar i == IRString "actor31SDF" -> createActorSDF context Actor31 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor32SDF" -> createActorSDF context Actor32 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor32SDF" -> createActorSDF context Actor32 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor32SDF" -> createActorSDF context Actor32 binder expr
+  -- 3 signals
   Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _)))
     | IRVar i == IRString "actor32SDF" -> createActorSDF context Actor32 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor33SDF" -> createActorSDF context Actor33 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor33SDF" -> createActorSDF context Actor33 binder expr
+  -- 2 signals
+  Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor33SDF" -> createActorSDF context Actor33 binder expr
+  -- 3 signals
   Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _)))
     | IRVar i == IRString "actor33SDF" -> createActorSDF context Actor33 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor34SDF" -> createActorSDF context Actor34 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor34SDF" -> createActorSDF context Actor34 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor34SDF" -> createActorSDF context Actor34 binder expr
+  -- 3 signals
   Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _)))
     | IRVar i == IRString "actor34SDF" -> createActorSDF context Actor34 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor41SDF" -> createActorSDF context Actor41 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor41SDF" -> createActorSDF context Actor41 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor41SDF" -> createActorSDF context Actor41 binder expr
+  -- 3 signals
+  Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _)))
+    | IRVar i == IRString "actor41SDF" -> createActorSDF context Actor41 binder expr
+  -- 4 signals
   Lam _ (Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _))))
     | IRVar i == IRString "actor41SDF" -> createActorSDF context Actor41 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor42SDF" -> createActorSDF context Actor42 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor42SDF" -> createActorSDF context Actor42 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor42SDF" -> createActorSDF context Actor42 binder expr
+  -- 3 signals
+  Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _)))
+    | IRVar i == IRString "actor42SDF" -> createActorSDF context Actor42 binder expr
+  -- 4 signals
   Lam _ (Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _))))
     | IRVar i == IRString "actor42SDF" -> createActorSDF context Actor42 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor43SDF" -> createActorSDF context Actor43 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor43SDF" -> createActorSDF context Actor43 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor43SDF" -> createActorSDF context Actor43 binder expr
+  -- 3 signals
+  Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _)))
+    | IRVar i == IRString "actor43SDF" -> createActorSDF context Actor43 binder expr
+  -- 4 signals
   Lam _ (Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _) _))))
     | IRVar i == IRString "actor43SDF" -> createActorSDF context Actor43 binder expr
+  -- 0 signals
+  App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _
+    | IRVar i == IRString "actor44SDF" -> createActorSDF context Actor44 binder expr
+  -- 1 signal
+  Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _)
+    | IRVar i == IRString "actor44SDF" -> createActorSDF context Actor44 binder expr
+  -- 2 signals
+  Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _))
+    | IRVar i == IRString "actor44SDF" -> createActorSDF context Actor44 binder expr
+  -- 3 signals
+  Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _) _)))
+    | IRVar i == IRString "actor44SDF" -> createActorSDF context Actor44 binder expr
+  -- 4 signals
   Lam _ (Lam _ (Lam _ (Lam _ (App (App (App (App (App (App (App (App (App (App (App (App (App (App (App (Var i) _) _) _) _) _) _) _) _) _) _) _) _) _) _) _))))
     | IRVar i == IRString "actor44SDF" -> createActorSDF context Actor44 binder expr
   _ -> createFunction context binder expr
</pre>
</details>

sthaeron committed:
CoreIRToProceduralIR: support eta-reduce

- strip type related Lambdas
- changes FunctionContent to only propagate variables which are
  associated with a input expression
- updates SDF example 27 to have delay defined outside of system

sthaeron commented:
> Tried this example and for me it doesn't actually compile (the type coercions are still in the functions). But we can get to ForSyDeIR at least, which is nice for the LSP.

That's my bad, forgot to mention I had to make a change to the Core IR to Procedural IR translation to make it work. Nevertheless, I've added a commit that would let us compile any function without type definitions. We can now compile SDF example 27 into C code; I had to make a slight change to the example since we do not support defining process constructors within the system netlist.

klarasm commented on examples/model/SDF_example_027.hs:
```diff
@@ -34,7 +36,7 @@ p_5 = actor11SDF 1 1 f_5
 
 -- Test Input
-s_in = signal [1 .. 10]
+-- s_in = signal [1 .. 10]
```
You don't actually need to comment this out since we ignore unused functions in this PR

klarasm commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -196,15 +204,15 @@ translateOutputExprToFunctionContentList context expr acc = case expr of
+    -- Only adds current variable id if its associated to a current variable expression
+    _ -> case lookup (IRVar varId) (functionVariables context) of
+      Just _ -> FVar varId : acc
+      Nothing -> acc
```
Nice! This makes it easier to understand.

klarasm commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -71,10 +70,19 @@ translateIRFunction function dflags constructors =
+stripLamForFunctionExpr :: CoreExpr -> CoreExpr
+stripLamForFunctionExpr expr = case expr of
+  Lam b e | isTyVar b -> stripLamForFunctionExpr e
+  Lam b e | isTcTyVar b -> stripLamForFunctionExpr e
```
This match is included in the first (see GHC compiler/GHC/Types/Var.h:1298, so I'll just remove it.

sthaeron replied:
```haskell
isTyVar :: Var -> Bool        -- True of both TyVar and TcTyVar
isTyVar (TyVar {})   = True
isTyVar (TcTyVar {}) = True
isTyVar _            = False
```
Oh, you are right, it is.

sthaeron committed:
CoreIRToForSyDeIR: support type-patterns in translateSystemExpr

sthaeron committed:
exmaples: add applications of processes using inferred types

sthaeron commented:
@klarasm seems like we have been working on solving the stripping consequences in parallel. I tested your new solution before committing mine. I found that it still had problems being able to compile examples where a process constructor did not use a type and was non-eta reduced. I've added SDF examples 28 and 29 to better help test the eta reduced cases.

The main difference between my solution and yours is the inclusion of not counting the stripped Lambda, which relates to the inferred type. For some reason, it is a GHC Var id rather than a type. I also added identifying a function within a where scope as a pattern-matching case rather than a separate function, mainly for readability.

klarasm commented:
> @klarasm seems like we have been working on solving the stripping consequences in parallel. I tested your new solution before committing mine. I found that it still had problems being able to compile examples where a process constructor did not use a type and was non-eta reduced. I've added SDF examples 28 and 29 to better help test the eta reduced cases.
> 
> The main difference between my solution and yours is the inclusion of not counting the stripped Lambda, which relates to the inferred type. For some reason, it is a GHC Var id rather than a type. I also added identifying a function within a where scope as a pattern-matching case rather than a separate function, mainly for readability.

Yes, I also found another solution:
```Haskell
stripLams :: Integer -> CoreExpr -> (Integer, CoreExpr)
stripLams n expr = case expr of
  Lam b (Lam _ e) | isTyVar b -> stripLams n e
  Lam _ e | otherwise -> stripLams (n + 1) e
  _ -> (n, expr)
```
which works without comparing the strings. I'll cherry-pick your examples and verify that my version works with them too.

sthaeron commented on src/CoreIRToForSyDeIR.hs:
```diff
@@ -224,47 +225,27 @@ translateSystemExpr :: TranslationContext -> CoreExpr -> (Binder, TranslationCon
-        context1 = createSignalsFromArguments initialContext pcId arguments
-        binder = PcId pcId
-     in (binder, context1)
+    (PcId $ IRVar i, createSignalsFromArguments initialContext (IRVar i) [IRVar a])
```
I was thinking of making this change myself, but found it to be much less readable. We want our code to be simple and maintainable. Having local temporary variables makes things very clear.

klarasm replied:
Sure, I'll use your version for this instead.

klarasm commented:
> > @klarasm seems like we have been working on solving the stripping consequences in parallel. I tested your new solution before committing mine. I found that it still had problems being able to compile examples where a process constructor did not use a type and was non-eta reduced. I've added SDF examples 28 and 29 to better help test the eta reduced cases.
> > The main difference between my solution and yours is the inclusion of not counting the stripped Lambda, which relates to the inferred type. For some reason, it is a GHC Var id rather than a type. I also added identifying a function within a where scope as a pattern-matching case rather than a separate function, mainly for readability.
> 
> Yes, I also found another solution:
> 
> ```haskell
> stripLams :: Integer -> CoreExpr -> (Integer, CoreExpr)
> stripLams n expr = case expr of
>   Lam b (Lam _ e) | isTyVar b -> stripLams n e
>   Lam _ e | otherwise -> stripLams (n + 1) e
>   _ -> (n, expr)
> ```
> 
> which works without comparing the strings. I'll cherry-pick your examples and verify that my version works with them too.

This would probably work also in `CoreIRToProceduralIR`, I'll use this approach there too if you have no objections @sthaeron?

```diff
diff --git a/src/CoreIRToProceduralIR.hs b/src/CoreIRToProceduralIR.hs
index f5693d0f6736..dfc1a3701b31 100644
--- a/src/CoreIRToProceduralIR.hs
+++ b/src/CoreIRToProceduralIR.hs
@@ -78,8 +78,7 @@ translateIRFunction function dflags constructors =
 -- variables.
 stripLamForFunctionExpr :: CoreExpr -> CoreExpr
 stripLamForFunctionExpr expr = case expr of
-  Lam b e | isTyVar b -> stripLamForFunctionExpr e
-  Lam b e | isId b -> if "ds_" `isPrefixOf` showPprUnsafe b then expr else stripLamForFunctionExpr e
+  Lam b (Lam _ e) | isTyVar b -> stripLamForFunctionExpr e
   _ -> expr
 
 -- Helper function which returns the `IRActor`s associated with a provided
```

sthaeron commented on src/CoreIRToForSyDeIR.hs:
```diff
@@ -349,6 +330,29 @@ getSourceFromArgument context argument =
+
+-- Strip internal Lets, likely from where scopes. Add any NonRecs as functions
+stripLets :: TranslationContext -> CoreExpr -> (TranslationContext, CoreExpr)
+stripLets context expr = case expr of
```
I would suggest not making this a separate function and integrating it as a pattern-matching statement in `translateCoreExpr`; I found that it made the code a lot more understandable.

sthaeron commented:
> This would probably work also in `CoreIRToProceduralIR`

I had just been testing in on my local branch and it works, yes!

klarasm commented:
I think this should now address your comments.

klarasm commented:
Hmm:

compiler/GHC/Types/Var.hs:1250
```Haskell
-- | Is this a coercion variable?
-- Satisfies @'isId' v ==> 'isCoVar' v == not ('isNonCoVarId' v)@.
isCoVar :: Var -> Bool
isCoVar (Id { id_details = details }) = isCoVarDetails details
isCoVar _                             = False
```

~This might be what we would use instead of `Lam b (Lam _ e) | isTyVar b ->` (or actually `isTyCoVar` which matches both).~

Nvm. This breaks weirdly.

klarasm commented:
WTF, now I just get:
```
#define PLATFORM PC
typedef int token;
#include "include/common.h"
#include <stdio.h>

forsyde-compiler-exe: <<loop>>
```
and exits. What have I done now...

klarasm commented:
Hm. I was a bit careless when inlining `stripLets`:
```diff
diff --git a/src/CoreIRToForSyDeIR.hs b/src/CoreIRToForSyDeIR.hs
index 0f50058aa129..b1f9aead5d55 100644
--- a/src/CoreIRToForSyDeIR.hs
+++ b/src/CoreIRToForSyDeIR.hs
@@ -412,8 +412,8 @@ translateCoreExpr context' binder expr' =
   let (n, expr1) = stripLams 0 expr'
       (context, expr2) = case expr1 of
         Let (NonRec bn be) e -> (createFunction context' bn be, e)
-        Let _ e -> (context, e)
-        _ -> (context, expr)
+        Let _ e -> (context', e)
+        _ -> (context', expr1)
       expr = maybe expr' id (stripApps n expr2)
    in case expr of
         App (App (Var i) _) _
```
makes it work again. ~Let me also try `isTyCoVar` without this fudged up inlining~ (it breaks SDF_example_028.hs, but the Delay.hs works (is just an inf-loop)...).

What wouldn't I give for explicit recursion sometimes...

sthaeron committed:
examples: add new combination of typed/untyped definitions

Add SDF example 30 which tests systems netlists with an explict type
declaration also includes different configuration of typed and untyped,
untyped, eta reduced and non-eta reduced

sthaeron committed:
CoreIRToForSyDeIR: ignore typed Lambdas in system netlist translation

sthaeron commented:
@klarasm While re-testing the PR today, I noticed that we are fully supporting system netlist which are not explicitly typed. Turns out SDF example 27 was a special case when compiling to Core IR since it only has one global system input. But for SDF example 28, when removing the system netlist type declration you would get the following as system inputs in ForSyDe IR `{"c", "$dNum", "s_in_x", "s_in_y"}`. Nevertheless, I've added a commit which fixes this problem and included a new SDF example 30 which tests it.

klarasm commented:
> @klarasm While re-testing the PR today, I noticed that we are fully supporting system netlist which are not explicitly typed. Turns out SDF example 27 was a special case when compiling to Core IR since it only has one global system input. But for SDF example 28, when removing the system netlist type declration you would get the following as system inputs in ForSyDe IR `{"c", "$dNum", "s_in_x", "s_in_y"}`. Nevertheless, I've added a commit which fixes this problem and included a new SDF example 30 which tests it.

Nice! I'll squash this into your earlier commit except the new example.

sthaeron reviewed (approved):
Except for my additional commit to ignore typed Lambdas in the system netlist, all implemented features are working and passing tests. Feel free to do any final rebasing if you feel it's required. Otherwise, this PR is ready to be merged and has my approval.
